projects/graph/src/draw.py

```diff
@@ -1,14 +1,75 @@
-"""
-General drawing methods for graphs using Bokeh.
-"""
+import math
 
 from bokeh.io import show, output_file
 from bokeh.plotting import figure
-from bokeh.models import (GraphRenderer, StaticLayoutProvider, Circle, LabelSet,
-                          ColumnDataSource)
+from bokeh.models import GraphRenderer, StaticLayoutProvider, Oval, Rect
+from bokeh.palettes import Spectral8
+from graph import Graph
 
 
 class BokehGraph:
     """Class that takes a graph and exposes drawing methods."""
-    def __init__(self):
-        pass  # TODO
+
+    def __init__(self, graph):
+        self.graph = graph
+
+    def draw(self):
+        graph = self.graph
+        N = len(graph.vertices)
+        node_indices = list(graph.vertices.keys())
+
+        print(node_indices)
+
+        plot = figure(title='Graph Layout Demonstration', x_range=(-7, 7), y_range=(-7, 7),
+                      tools='', toolbar_location=None)
+
+        graph_renderer = GraphRenderer()
+
+        graph_renderer.node_renderer.data_source.add(node_indices, 'index')
+        node_colors = ['red', 'blue', 'green', 'yellow',
+                       'black', 'pink', 'purple', 'white']
+        graph_renderer.node_renderer.data_source.add(node_colors, 'color')
+        graph_renderer.node_renderer.glyph = Rect(
+            height=0.1, width=0.2, fill_color='color')
+
+        graph_renderer.edge_renderer.data_source.data = dict(
+            start=[0]*N,
+            end=node_indices)
+
+        d = dict(
+            start=[0]*N,
+            end=node_indices)
+        print(d)
+
+        # start of layout code
+        # circ = [i*2*math.pi/8 for i in node_indices]
+        # x = [math.cos(i) for i in circ]
+        # y = [math.sin(i) for i in circ]
+
+        x = []
+        y = []
+        for vertex_id in node_indices:
+            vertex = graph.vertices[vertex_id]
+            x.append(vertex.x)
+            y.append(vertex.y)
+
+        graph_layout = dict(zip(node_indices, zip(x, y)))
+        graph_renderer.layout_provider = StaticLayoutProvider(
+            graph_layout=graph_layout)
+
+        plot.renderers.append(graph_renderer)
+
+        output_file('graph.html')
+        show(plot)
+
+
+graph = Graph()  # Instantiate your graph
+graph.add_vertex('0')
+graph.add_vertex('1')
+graph.add_vertex('2')
+graph.add_vertex('3')
+graph.add_edge('0', '1')
+graph.add_edge('0', '3')
+
+bg = BokehGraph(graph)
+bg.draw()
```

Comment:
I think there's something about using Spectral8 that throws a warning, because I'm getting a message that says color is 8 but index is 4 and that these two numbers have to match up. Just so you know. This is a great template for getting started but you may want to look in the docs for a different color scheme/randomizer.